File: cpp/graph/dfs.cpp
#include "dfs.h"

void dfsVisit(Graph const &graph, int u, vector<int> &pred,
              vector<vertexColor> &color) {
  color[u] = Gray;

  for (VertexList::const_iterator edge_iter = graph.begin(u);
       edge_iter != graph.end(u); ++edge_iter) {
    int v = edge_iter->first;

    if (color[v] == White) {
      pred[v] = u;
      dfsVisit(graph, v, pred, color);
    }
  }

  color[u] = Black;
}

void dfsSearch(Graph const &graph, int s, vector<int> &pred) {
  const int n = graph.numVertices();
  vector<vertexColor> color(n, White);
  pred.assign(n, -1);

  dfsVisit(graph, s, pred, color);

  for (int u = 0; u < n; u++) {
    if (color[u] == White) {
      dfsVisit(graph, u, pred, color);
    }
  }
}

File: typescript/graph/dfs.ts
import Graph from "./graph.ts";

const vertexColor = ["White", "Gray", "Black"] as const;
type VertexColor = typeof vertexColor[number];

const dfsVisit = (
  graph: Graph,
  u: number,
  pred: number[],
  color: VertexColor[]
) => {
  color[u] = "Gray";

  for (const [v, w] of graph.getVertices(u)) {
    if (color[v] === "White") {
      pred[v] = u;
      dfsVisit(graph, v, pred, color);
    }
  }

  color[u] = "Black";
};

const dfsSearch = (graph: Graph, start: number): number[] => {
  const n = graph.getVerticesNum();
  const color: VertexColor[] = Array(n).fill("White");
  const pred: number[] = Array(n).fill(-1);

  dfsVisit(graph, start, pred, color);

  for (let u = 0; u < n; u++) {
    if (color[u] == "White") {
      dfsVisit(graph, u, pred, color);
    }
  }

  return pred;
};

export default dfsSearch;
